File: src/main/java/xyz/emirdev/echogen/commands/VanishCommand.java
package xyz.emirdev.echogen.commands;

import org.bukkit.entity.Player;

import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.tree.LiteralCommandNode;

import io.papermc.paper.command.brigadier.CommandSourceStack;
import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.command.brigadier.argument.ArgumentTypes;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import xyz.emirdev.echogen.Echogen;
import xyz.emirdev.echogen.PluginCommand;

public class VanishCommand extends PluginCommand {

    @Override
    public LiteralCommandNode<CommandSourceStack> getCommand() {
        return Commands.literal("vanish")
                .requires(hasPermission("echogen.vanish"))
                .executes(this::vanish)
                .then(Commands.argument("player", ArgumentTypes.player())
                        .requires(hasPermission("echogen.vanish.others"))
                        .executes(this::vanishOther))
                .build();
    }

    private int vanish(CommandContext<CommandSourceStack> ctx) {
        Player player = getContextPlayer(ctx);
        if (player == null)
            return 1;

        boolean vanished = Echogen.get().getVanishManager().isVanished(player);

        if (!vanished) {
            Echogen.get().getVanishManager().vanish(player);
            player.sendRichMessage("<aqua>You have <green>enabled</green> vanish.</aqua>");
        } else {
            Echogen.get().getVanishManager().unVanish(player);
            player.sendRichMessage("<aqua>You have <red>disabled</red> vanish.</aqua>");
        }

        return 1;
    }

    private int vanishOther(CommandContext<CommandSourceStack> ctx) throws CommandSyntaxException {
        Player player = getContextPlayer(ctx);
        if (player == null)
            return 1;

        Player target = getPlayer(ctx, "player");

        boolean vanished = Echogen.get().getVanishManager().isVanished(target);

        TagResolver playerTag = Placeholder.unparsed("player", target.getName());

        if (!vanished) {
            Echogen.get().getVanishManager().vanish(target);
            player.sendRichMessage("<aqua>You have <green>enabled</green> vanish for <player>.</aqua>", playerTag);
            target.sendRichMessage("<aqua>Your vanish has been <green>enabled</green>.</aqua>");
        } else {
            Echogen.get().getVanishManager().unVanish(target);
            player.sendRichMessage("<aqua>You have <red>disabled</red> vanish for <player>.</aqua>", playerTag);
            target.sendRichMessage("<aqua>Your vanish has been <red>disabled</red>.</aqua>");
        }

        return 1;
    }
}

File: src/main/kotlin/cat/emir/echogen/commands/VanishCommand.kt
package cat.emir.echogen.commands

import cat.emir.echogen.Echogen
import com.mojang.brigadier.context.CommandContext

import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.argument.ArgumentTypes
import cat.emir.echolib.command.PluginCommand
import cat.emir.echolib.command.getPlayer
import cat.emir.echolib.sendLangMessage
import com.mojang.brigadier.builder.LiteralArgumentBuilder

class VanishCommand(plugin: Echogen) : PluginCommand<Echogen>(plugin) {

    override fun getCommand(): LiteralArgumentBuilder<CommandSourceStack> {
        return command("vanish") {
            requires { it.sender.hasPermission("echogen.vanish") }
            executes(::vanish)
            argument("player", ArgumentTypes.player()) {
                requires { it.sender.hasPermission("echogen.vanish.others") }
                executes(::vanishOther)
            }
        }
    }

    fun vanish(ctx: CommandContext<CommandSourceStack>): Int {
        val player = ctx.getPlayer() ?: return 1
        val vanished = plugin.vanishManager.isVanished(player)

        if (!vanished) {
            plugin.vanishManager.vanish(player)
            player.sendLangMessage("vanish.self.enabled")
        } else {
            plugin.vanishManager.unVanish(player)
            player.sendLangMessage("vanish.self.disabled")
        }

        return 1
    }

    fun vanishOther(ctx: CommandContext<CommandSourceStack>): Int {
        val sender = ctx.source.sender
        val target = ctx.getPlayer("player") ?: return 1
        val vanished = plugin.vanishManager.isVanished(target)

        val data = listOf("player" to target.name)

        if (!vanished) {
            plugin.vanishManager.vanish(target)
            sender.sendLangMessage("vanish.other.enabled.executor", data)
            target.sendLangMessage("vanish.other.enabled.target")
        } else {
            plugin.vanishManager.unVanish(target)
            sender.sendLangMessage("vanish.other.disabled.executor", data)
            target.sendLangMessage("vanish.other.disabled.target")
        }

        return 1
    }
}
